Slackware service state initial support

piterpunk opened the pull request:
### What does this PR do?
Adds support to manage services in Slackware Linux.

Without this module even a simple operation as service.start doesn't works, that makes impossible to use any service state.

Now, with this module, all service module operations works fine (provided the managed rc-scripts provides the required operation) and the service states are available.

### What issues does this PR fix or reference?
Fixes: #58206 

### Previous Behavior
* Calling a service.start:
```
# salt-call service.start ntp
Error running 'service.start': Unable to run command '['/etc/init.d/ntp', 'start']' with the context '{'cwd': '/root', 'shell': False, 'env': {'SHELL': '/bin/bash', 'LESS': '-M', 'PKG_CONFIG_PATH': '/usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig', 'GNOME_KEYRING_CONTROL': '/root/.cache/keyring-WUK8O0', 'G_BROKEN_FILENAMES': '1', 'HOSTNAME': 'marvin.mylab', 'MINICOM': '-c on', 'QT4DIR': '/usr/lib/qt', 'PWD': '/srv/salt/ntp', 'LOGNAME': 'root', 'LS_OPTIONS': '-F -b -T 0 --color=auto', 'HOME': '/root', 'LANG': 'en_US.UTF-8', 'LS_COLORS': 'no=00:fi=00:di=01;34:ln=01;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.bat=01;32:*.btm=01;32:*.cmd=01;32:*.com=01;32:*.dll=01;32:*.exe=01;32:*.7z=01;31:*.ace=01;31:*.arj=01;31:*.bz2=01;31:*.cpio=01;31:*.deb=01;31:*.dz=01;31:*.gz=01;31:*.jar=01;31:*.lha=01;31:*.lz=01;31:*.lzh=01;31:*.lzma=01;31:*.rar=01;31:*.rpm=01;31:*.rz=01;31:*.tar=01;31:*.taz=01;31:*.tb2=01;31:*.tbz2=01;31:*.tbz=01;31:*.tgz=01;31:*.tlz=01;31:*.trz=01;31:*.txz=01;31:*.tz=01;31:*.tz2=01;31:*.xz=01;31:*.z=01;31:*.zip=01;31:*.zoo=01;31:*.aac=01;35:*.anx=01;35:*.asf=01;35:*.au=01;35:*.axa=01;35:*.axv=01;35:*.avi=01;35:*.bmp=01;35:*.divx=01;35:*.flac=01;35:*.gif=01;35:*.ico=01;35:*.jpg=01;35:*.jpeg=01;35:*.m2a=01;35:*.m2v=01;35:*.m4a=01;35:*.m4p=01;35:*.m4v=01;35:*.mid=01;35:*.midi=01;35:*.mka=01;35:*.mkv=01;35:*.mov=01;35:*.mp3=01;35:*.mp4=01;35:*.mp4v=01;35:*.mpc=01;35:*.mpeg=01;35:*.mpg=01;35:*.nuv=01;35:*.oga=01;35:*.ogv=01;35:*.ogx=01;35:*.ogg=01;35:*.opus=01;35:*.pbm=01;35:*.pgm=01;35:*.png=01;35:*.ppm=01;35:*.qt=01;35:*.ra=01;35:*.ram=01;35:*.rm=01;35:*.spx=01;35:*.svg=01;35:*.svgz=01;35:*.tga=01;35:*.tif=01;35:*.tiff=01;35:*.vob=01;35:*.wav=01;35:*.wma=01;35:*.wmv=01;35:*.xbm=01;35:*.xcf=01;35:*.xpm=01;35:*.xspf=01;35:*.xwd=01;35:*.xvid=01;35:', 'SSH_CONNECTION': '192.168.0.19 45190 192.168.0.42 22', 'TERM': 'xterm-256color', 'G_FILENAME_ENCODING': '@locale', 'CPLUS_INCLUDE_PATH': '/usr/lib/qt/include', 'LESSOPEN': '|lesspipe.sh %s', 'USER': 'root', 'T1LIB_CONFIG': '/usr/share/t1lib/t1lib.config', 'GDK_USE_XFT': '1', 'SHLVL': '1', 'INPUTRC': '/etc/inputrc', 'PS2': '> ', 'PS1': '\\u@\\h:\\w\\$ ', 'SSH_CLIENT': '192.168.0.19 45190 22', 'QT5DIR': '/usr/lib/qt5', 'LC_COLLATE': 'C', 'VDPAU_LOG': '0', 'PATH': '/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/lib/qt/bin:/usr/lib/qt5/bin', 'SSH_TTY': '/dev/pts/1', '_': '/usr/bin/salt-call', 'OLDPWD': '/srv/salt', 'LC_CTYPE': 'C', 'LC_NUMERIC': 'C', 'LC_TIME': 'C', 'LC_MONETARY': 'C', 'LC_MESSAGES': 'C', 'LC_PAPER': 'C', 'LC_NAME': 'C', 'LC_ADDRESS': 'C', 'LC_TELEPHONE': 'C', 'LC_MEASUREMENT': 'C', 'LC_IDENTIFICATION': 'C', 'LANGUAGE': 'C'}, 'stdin': None, 'stdout': -1, 'stderr': -2, 'with_communicate': True, 'timeout': None, 'bg': False, 'close_fds': True}', reason: [Errno 2] No such file or directory: '/etc/init.d/ntp'
```
* Trying to apply service.running state:
```
          ID: ntp_service-running-ntpd
    Function: service.running
        Name: ntpd
      Result: False
     Comment: Unable to run command '['/etc/init.d/ntpd', 'restart']' with the context '{'cwd': '/root', 'shell': False, 'env': {'SHELL': '/bin/bash', 'LESS': '-M', 'PKG_CONFIG_PATH': '/usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig', 'GNOME_KEYRING_CONTROL': '/root/.cache/keyring-WUK8O0', 'G_BROKEN_FILENAMES': '1', 'HOSTNAME': 'marvin.mylab', 'MINICOM': '-c on', 'QT4DIR': '/usr/lib/qt', 'PWD': '/srv/salt/ntp', 'LOGNAME': 'root', 'LS_OPTIONS': '-F -b -T 0 --color=auto', 'HOME': '/root', 'LANG': 'en_US.UTF-8', 'LS_COLORS': 'no=00:fi=00:di=01;34:ln=01;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.bat=01;32:*.btm=01;32:*.cmd=01;32:*.com=01;32:*.dll=01;32:*.exe=01;32:*.7z=01;31:*.ace=01;31:*.arj=01;31:*.bz2=01;31:*.cpio=01;31:*.deb=01;31:*.dz=01;31:*.gz=01;31:*.jar=01;31:*.lha=01;31:*.lz=01;31:*.lzh=01;31:*.lzma=01;31:*.rar=01;31:*.rpm=01;31:*.rz=01;31:*.tar=01;31:*.taz=01;31:*.tb2=01;31:*.tbz2=01;31:*.tbz=01;31:*.tgz=01;31:*.tlz=01;31:*.trz=01;31:*.txz=01;31:*.tz=01;31:*.tz2=01;31:*.xz=01;31:*.z=01;31:*.zip=01;31:*.zoo=01;31:*.aac=01;35:*.anx=01;35:*.asf=01;35:*.au=01;35:*.axa=01;35:*.axv=01;35:*.avi=01;35:*.bmp=01;35:*.divx=01;35:*.flac=01;35:*.gif=01;35:*.ico=01;35:*.jpg=01;35:*.jpeg=01;35:*.m2a=01;35:*.m2v=01;35:*.m4a=01;35:*.m4p=01;35:*.m4v=01;35:*.mid=01;35:*.midi=01;35:*.mka=01;35:*.mkv=01;35:*.mov=01;35:*.mp3=01;35:*.mp4=01;35:*.mp4v=01;35:*.mpc=01;35:*.mpeg=01;35:*.mpg=01;35:*.nuv=01;35:*.oga=01;35:*.ogv=01;35:*.ogx=01;35:*.ogg=01;35:*.opus=01;35:*.pbm=01;35:*.pgm=01;35:*.png=01;35:*.ppm=01;35:*.qt=01;35:*.ra=01;35:*.ram=01;35:*.rm=01;35:*.spx=01;35:*.svg=01;35:*.svgz=01;35:*.tga=01;35:*.tif=01;35:*.tiff=01;35:*.vob=01;35:*.wav=01;35:*.wma=01;35:*.wmv=01;35:*.xbm=01;35:*.xcf=01;35:*.xpm=01;35:*.xspf=01;35:*.xwd=01;35:*.xvid=01;35:', 'SSH_CONNECTION': '192.168.0.19 45190 192.168.0.42 22', 'TERM': 'xterm-256color', 'G_FILENAME_ENCODING': '@locale', 'CPLUS_INCLUDE_PATH': '/usr/lib/qt/include', 'LESSOPEN': '|lesspipe.sh %s', 'USER': 'root', 'T1LIB_CONFIG': '/usr/share/t1lib/t1lib.config', 'GDK_USE_XFT': '1', 'SHLVL': '1', 'INPUTRC': '/etc/inputrc', 'PS2': '> ', 'PS1': '\\u@\\h:\\w\\$ ', 'SSH_CLIENT': '192.168.0.19 45190 22', 'QT5DIR': '/usr/lib/qt5', 'LC_COLLATE': 'C', 'VDPAU_LOG': '0', 'PATH': '/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/lib/qt/bin:/usr/lib/qt5/bin', 'SSH_TTY': '/dev/pts/1', '_': '/usr/bin/salt-call', 'OLDPWD': '/srv/salt', 'LC_CTYPE': 'C', 'LC_NUMERIC': 'C', 'LC_TIME': 'C', 'LC_MONETARY': 'C', 'LC_MESSAGES': 'C', 'LC_PAPER': 'C', 'LC_NAME': 'C', 'LC_ADDRESS': 'C', 'LC_TELEPHONE': 'C', 'LC_MEASUREMENT': 'C', 'LC_IDENTIFICATION': 'C', 'LANGUAGE': 'C'}, 'stdin': None, 'stdout': -1, 'stderr': -2, 'with_communicate': True, 'timeout': None, 'bg': False, 'close_fds': True}', reason: [Errno 2] No such file or directory: '/etc/init.d/ntpd'
     Started: 19:41:35.631266
    Duration: 21163.166 ms
     Changes:   
```


### New Behavior
* Calling a service.start
```
# salt-call service.start ntpd
local:
    True
```

* Applying a service.running state:
```
          ID: ntp_service-running-ntpd
    Function: service.running
        Name: ntpd
      Result: True
     Comment: The service ntpd is already running
     Started: 19:48:23.647516
    Duration: 21.559 ms
     Changes:
```   

### Merge requirements satisfied?
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Commit:
Initial support to manage services in Slackware

- Added slackware_service.py module to handle the Slackware rc-scripts
- Added Slackware in the list to not use the linux_service.py

Commit:
Added unit tests to slackware_service.py module

- Added tests/init/modules/test_slackware_service.py
- Fixed the permission mask on slackware_service.py
- Added slackware_service in index.rst

Commit:
Added Changelog entry for #58206

Commit:
Fixed slackware_service's enable and disable tests

- Call to os.stat wasn't properly mocked and the tests fails on all
  non-Slackware platforms.
- Fixed the affected tests and changed the test argument to a
  non-existent file, so the wrong mock can be detected in Slackware.

Commit:
Fixed underlines in RST file

piterpunk commented:
@dwoz , can you help on this? All the builds failed without actually do any build:
```
09:53:01  Running on EC2 (saltstack-ci) - kitchen (i-075b7ee69471a539c) in /var/jenkins/workspace/pr-fedora32-py3/PR-58207
09:53:01  [Pipeline] {
09:53:01  [Pipeline] timeout
09:53:01  Timeout set to expire in 7 hr 0 min
09:53:01  [Pipeline] {
09:53:01  [Pipeline] withEnv
09:53:01  [Pipeline] {
09:53:01  [Pipeline] stage
09:53:01  [Pipeline] { (Clone)
09:53:01  [Pipeline] cleanWs
09:53:01  [Pipeline] }
09:53:02  [Pipeline] // stage
09:53:02  [Pipeline] }
09:53:02  [Pipeline] // withEnv
09:53:02  [Pipeline] cleanWs
09:53:02  [Pipeline] }
09:53:02  [Pipeline] // timeout
09:53:02  [Pipeline] }
09:53:02  [Pipeline] // node
09:53:02  [Pipeline] echo
09:53:02  Setting currentBuild.result to FAILURE
```
I don't know what happened :(

Ch3LL commented on tests/unit/modules/test_slackware_service.py:
```diff
@@ -0,0 +1,234 @@
+"""
```
we are still in the process of migrating entirely to pytest, but since this is a new test file i would love to see it using pytest instead in `tests/pytests/unit/modules` so we dont have to migrate it later if you're willing to do so.

piterpunk replied:
Oh, ok. I will change it, as it's a new feature it's better to born in the right way.

Ch3LL replied:
thanks!

piterpunk replied:
Ok, I guess it's done.

piterpunk commented on salt/modules/slackware_service.py:
```diff
@@ -0,0 +1,328 @@
+
+    .. code-block:: bash
+
+        salt '*' service.start <service name>
```
OK, I will do that tonight

piterpunk replied:
Done

piterpunk commented on salt/modules/slackware_service.py:
```diff
@@ -0,0 +1,328 @@
+# Define the module's virtual name
+__virtualname__ = "service"
+
+prefix = "/etc/rc.d/rc"
```
Yes, this is always the path for rc scripts in Slackware. It's like /etc/init.d on SysV init.

piterpunk commented on salt/modules/slackware_service.py:
```diff
@@ -0,0 +1,328 @@
+    If the name contains globbing, a dict mapping service name to True/False
+    values is returned.
+
+    .. versionchanged:: 2018.3.0
```
Hahahaha... sorry, this what happens when copying and pasting stuff, I will fix that tonight.

piterpunk replied:
Done

Commit:
Migrated tests to PyTest and added release version

Ch3LL commented on tests/pytests/unit/modules/test_slackware_service.py:
```diff
@@ -0,0 +1,251 @@
+# Import Salt Testing Libs
+from tests.support.mock import MagicMock, patch
+
+glob_output = [
```
thanks for all these updates, really appreciate them. I have just one more thing. Can we add these variables to a pytest fixture? This actually helps speed up things when pytest is collecting tests and prevents any issues we have ran into the past where someone edited a list/dict at the global module level that affects other tests.

piterpunk replied:
It took a while, but following your advice and with the explanations of the @waynew from the Saltstack Test Clinic, I removed all the global variables and took the opportunity to make a big overhaul of test names, descriptions, comments and reorganized some them. I hope this time everything is fine :)

Commit:
Many changes in slackware_service unit tests

- Removed global variables and put inside a reusable fixture
- Changed the test names to be more descriptive
- Also added better explanations and comments to the mocked data and the
  tests themselves
- Changed the enabled/disabled tests to be more useful

Ch3LL reviewed (approved):
thanks for all your work on this one, much appreciated

piterpunk commented:
@dwoz why the "Needs Testcase" label?
